standalone/pruntime/enclave/src/system/mod.rs

```diff
@@ -487,67 +528,112 @@ impl System {
+                hex::encode(master_key.public())
+            );
+            let master_pubkey = RegistryEvent::MasterPubkey {
+                master_pubkey: master_key.public(),
+            };
+            self.egress.send(&master_pubkey);
+        }
 
-                info!("Init gatekeeper in block {}", block.block_number);
-                let gatekeeper = gk::Gatekeeper::new(
-                    master_key.clone(),
-                    &mut block.recv_mq,
-                    self.send_mq
-                        .channel(MessageOrigin::Gatekeeper, master_key.clone()),
-                );
-                self.gatekeeper = Some(gatekeeper);
-            }
-        } else {
-            if let Some(gatekeeper) = &mut self.gatekeeper {
-                gatekeeper.share_master_key(&event.pubkey, &event.ecdh_pubkey);
-            }
+        if self.master_key.is_some() {
+            info!("Init gatekeeper in block {}", block.block_number);
+            self.init_gatekeeper(&mut block.recv_mq);
         }
 
         if my_pubkey == event.pubkey {
+            self.gatekeeper
+                .as_mut()
+                .expect("gatekeeper must be initializaed here; qed.")
+                .register_on_chain();
             self.registered_on_chain = true;
-            if let Some(gatekeeper) = &mut self.gatekeeper {
+        }
+    }
+
+    fn process_gatekeeper_change_event(
+        &mut self,
+        block: &mut BlockInfo,
+        origin: MessageOrigin,
+        event: GatekeeperChange,
+    ) {
+        info!("Incoming gatekeeper change event: {:?}", event);
+        match event {
+            GatekeeperChange::GatekeeperRegistered(new_gatekeeper_event) => {
+                self.process_new_gatekeeper_event(block, origin, new_gatekeeper_event)
+            }
+        }
+    }
+
+    /// Share the master key to the newly-registered gatekeeper
+    /// Tick the state if the registered gatekeeper is this worker
+    fn process_new_gatekeeper_event(
+        &mut self,
+        block: &mut BlockInfo,
+        origin: MessageOrigin,
+        event: NewGatekeeperEvent,
+    ) {
+        if !origin.is_pallet() {
+            error!("Invalid origin {:?} sent a {:?}", origin, event);
+            return;
+        }
+
+        // double check the registered gatekeeper is valid on chain
+        if !chain_state::is_gatekeeper(&event.pubkey, block.storage) {
+            error!(
+                "Fatal error: Invalid first gatekeeper registration {:?}",
+                event
+            );
+            panic!("System state poisoned");
+        }
+
+        if let Some(gatekeeper) = &mut self.gatekeeper {
+            gatekeeper.share_master_key(&event.pubkey, &event.ecdh_pubkey, block.block_number);
+
+            let my_pubkey = self.identity_key.public();
+            if my_pubkey == event.pubkey {
                 gatekeeper.register_on_chain();
+                self.registered_on_chain = true;
```

Comment:
`registered_on_chain` is confusing in the context of `System` (usually it refers to a worker). I think it's purely duplicated as we can read it from `self.gatekeeper()`. We'd better remove this field completely.

Reply:
No, we cannot read it from `self.gatekeeper()` unless we check whether the gatekeeper egress is in dummy mode, while it's reasonable to move it into gatekeeper and rewrite the `gatekeeper_status()`.